Specify behavior when following a hidden subtree via aria-labelledby …

jaragunde opened the pull request:
…or -describedby.

This describes more accurately the actual behavior of Firefox and Blink. There is a fundamental difference between those two yet, Firefox does not hidden aria-hidden nodes while Blink does. My preference would be to use aria-hidden to allow authors to explicitly hide content if they really want but I wouldn't push hard for it.

About WebKit, its behavior changes depending on how nodes were hidden, `visibility:hidden` and `display:none` seem to have different behaviors which suggests to have followed the innerText implementation, maybe sharing code.

jaragunde committed:
Specify behavior when following a hidden subtree via aria-labelledby or -describedby.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,32 @@ <h4>Computation steps</h4>
+            <li id="step2A">If the <code>current node</code> is <a class="termref">hidden</a> <strong>and is:
+            <ul>
+              <li>Not</strong> directly referenced by <code>aria-labelledby</code> or <code>aria-describedby</code></li>
+              <li>Nor part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal started in <a href="#step2B.ii.b">step 2B.ii.b<a>, where the target of that relation was hidden</li>
```
The line above is the only new text, actually, save for the new examples.

jaragunde commented:
TODO: specify the possibility to use `aria-hidden=true` to explicitly hide a bit of the hidden subtree.

jaragunde committed:
Editorial changes.

jaragunde committed:
Add clause about aria-hidden.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
+              <li id="step2A.i">If the <code>current node</code> is <strong>not</strong> <a href="https://html.spec.whatwg.org/multipage/rendering.html#being-rendered">being rendered</a> <strong>and is</strong>:
+                <ul>
+                  <li><strong>Not</strong> directly referenced by <code>aria-labelledby</code> or <code>aria-describedby</code>.</li>
+                  <li><strong>Nor</strong> part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal started in <a href="#step2B.ii.b">step 2B.ii.b<a>, where the target of that relation was hidden.</li>
```
The line above is new. The rest of the surrounding text is the same but it has been reformatted, creating a numbered list and some bullet points.

jaragunde replied:
In the end I condensed the first two bullet points. See the new commits.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
+                    <p>By default, <a class="termref">assistive technologies</a> do not relay hidden information, but an author can explicitly override that and include hidden text as part of the <a class="termref">accessible name</a> or <a class="termref">accessible description</a> by using <code>aria-labelledby</code> or <code>aria-describedby</code>. </p>
+                  </details></div>
+                  <div><details>
+                    <summary>Example:</summary>
```
These examples are new, too.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
+                    </pre>
+                  </details></div>
+                </li>
+              <li id="step2A.ii">If the <code>current node</code> has the <code>aria-hidden</code> property set to true, return the empty string.
```
This clause, with the example below, are the only remaining additions to the text.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
-              </details></div>
+            <li id="step2A">
+            <ol>
+              <li id="step2A.i">If the <code>current node</code> is <strong>not</strong> <a href="https://html.spec.whatwg.org/multipage/rendering.html#being-rendered">being rendered</a> <strong>and is</strong>:
```
In the line above, I'm replacing the use of "hidden" with "not being rendered", like James Nurthen had proposed [here](https://github.com/w3c/accname/pull/137/commits/584a503aa9e63576cdac2ba97eeb683e99b3c7c8).

aleventhal commented:
> …or -describedby.
> 
> This describes more accurately the actual behavior of Firefox and Blink. There is a fundamental difference between those two yet, Firefox does not hidden aria-hidden 

Nit: does not hidden -> does not hide

> nodes while Blink does. My preference would be to use aria-hidden to allow authors to explicitly hide content if they really > want but I wouldn't push hard for it.

We should mention there are a few possibilities:
1. Always remove aria-hidden content from any name calculation (even if pointed to by a relation). This is a relatively simply rule to explain and remember. It is not very helpful in the repair cases for aria-hidden, though, e.g. when someone puts aria-hidden on the body.
2. Use the same rules for aria-hidden as other types of hidden. IOW use if part of a subtree that was hidden at the root, for any type of hidden. This is probably the simplest to explain and remember because all types of hidden become equivalent. This does not solve the tooltip case where the author wants to consistently remove part of the tooltip from the name, and the main tooltip goes in and out of being rendered.
3. Only remove aria-hidden content in descendant subtrees of a labelledby or describedby, or if it's not in a labelledby/describedby calc. This solves the tooltip case, but is very hard to explain and remember.

After writing all of these out, I'm starting to get a preference for #2. I think the tooltip case with some hidden content is contrived. The author can just remove those nodes. It's not worth creating the mess of #3. And #1 has the issue for repair situations. WDYT?

> About WebKit, its behavior changes depending on how nodes were hidden, `visibility:hidden` and `display:none` seem to have different behaviors which suggests to have followed the innerText implementation, maybe sharing code.

We should add this line at the bottom:
Closes #57

aleventhal commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
-              </details></div>
+            <li id="step2A">
+            <ol>
+              <li id="step2A.i">If the <code>current node</code> is <strong>not</strong> <a href="https://html.spec.whatwg.org/multipage/rendering.html#being-rendered">being rendered</a> <strong>and is</strong>:
```
I clicked on the "not being rendered" link and i'm confused by the definition. And if I as a browser developer am confused, then certainly authors will be as well. Therefore, I think we need specific guidance on each hiding technique so that no one has to guess. Maybe other reviewers will have an idea.

For example, from reading the "not-being-renderered", is it clear with visibility:hidden is rendered or not? It takes up space, unlike visibility:collapse, so that if an <a href>text</a> element is visibility:hidden, there is a visible hole where it would have been. That means it's rendered, correct?

However, I thought that for AccName we treated visibility:hidden the same as display:none.

Also, what about content-visibility? Some content-visibility is rendered, some isn't, depending on the scroll position? Are we saying the label would change depending on whether something is scrolled onscreen or not? It's probably better to just say that content-visiblity:auto content is treated as visible.

My preference would be to simply list the specific hiding techniques that lead to this condition, so there are no questions. But I'm not sure whether other reviewers will balk at having a dependency on another spec.

jaragunde replied:
You are right about "not being rendered", I understood it as "a type of hidden that's not aria-hidden", but it's actually describing display:none but not visibility:hidden.

It's probably better to point to the [hidden definition](https://www.w3.org/TR/accname-1.2/#dfn-hidden) and modify it to state clearly to which kinds of hidden we want to consider (as it is now, it might or might not include aria-hidden).

jaragunde replied:
This is done, I'm back to reference "hidden" and adding clarification to that.

aleventhal commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
+                    <p>The following examples show the meaning of the clause "Not part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal started in <a href="#step2B.ii.b">step 2B.ii.b<a>, where the target of that relation was hidden".</p>
+                    <p>First, <code>element1</code>'s <a class="termref">accessible name</a> is "hello" because, although <code>element3</code> is hidden, it is part of an <code>aria-labelledby</code> traversal started in <code>element2</code>, which is hidden too.</p>
+                    <pre class="example highlight"><code>&lt;element1 id="el1" aria-labelledby="el2" /&gt;
+&lt;element2 id="el2" hidden&gt;
```
I don't think everyone knows that the hidden attribute is exactly the same as display:none. Maybe the example should use style="display:none" ?

jaragunde replied:
I considered the most "semantic" version of the attribute would be appropriate, also to state that we refer to multiple kinds of hidden. No strong opinion, though.

jaragunde replied:
I settled to `class=hidden`, which I think it's semantic and doesn't tie the example to a specific type of hidden.

aleventhal commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
+                    <summary>Example:</summary>
+                    <p>The following examples show the meaning of the clause "Not part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal started in <a href="#step2B.ii.b">step 2B.ii.b<a>, where the target of that relation was hidden".</p>
+                    <p>First, <code>element1</code>'s <a class="termref">accessible name</a> is "hello" because, although <code>element3</code> is hidden, it is part of an <code>aria-labelledby</code> traversal started in <code>element2</code>, which is hidden too.</p>
+                    <pre class="example highlight"><code>&lt;element1 id="el1" aria-labelledby="el2" /&gt;
```
People might be wondering if element1 and element3 allow a name. Should we add role="button" to be clear?

jaragunde replied:
Sounds good!

aleventhal commented on index.html:
```diff
@@ -249,11 +249,49 @@ <h4>Computation steps</h4>
+                    <summary>Example:</summary>
+                    <p>The following example shows the interaction of the clauses in step 2A.i and step 2A.ii:</p>
+                    <p><code>element1</code>'s <a class="termref">accessible name</a> is "hello". <code>element4</code> does not contribute to the accessible name because its  <code>aria-hidden</code> attribute is true.</p>
+                    <pre class="example highlight"><code>&lt;element1 id="el1" aria-labelledby="el2" /&gt;
```
What if the root of the label is aria-hidden?

jaragunde replied:
If the root of the label is aria-hidden, it would still contribute its contents. I tried to specify this behavior with:

> <li><strong>Nor</strong> part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal started in <a href="#step2B.ii.b">step 2B.ii.b<a>, where the target of that relation was hidden.</li>

All in all, we are always back to the same question. What does hidden actually mean?

jaragunde replied:
Hope to have clarified that in the new "hidden" definition. Resolving this thread of conversation.

aleventhal reviewed (commented):
Good start. I added some thoughts. LMK what you think.

jaragunde commented:
> We should mention there are a few possibilities:
> 
> 1. Always remove aria-hidden content from any name calculation...
> 2. Use the same rules for aria-hidden as other types of hidden...
> 3. Only remove aria-hidden content in descendant subtrees...

I'm changing my mind and I'd go with 2 right now. It should make the text easier to write and understand, and that's very important in an already complicated spec.

The kind of authors who care about the scenario we would solve in option 3 are probably already implementing it in a way that works with 2 (remove the node), because that's the way Firefox has always worked.

I'll give a try at how the text would look like.

jaragunde committed:
Revert "Add clause about aria-hidden."

This reverts commit 83229c02468f2fca1a91898b0867852745ea2e13.

jaragunde committed:
Simplify cases in step 2A and add clarification.

jaragunde committed:
Add original hidden definition from wai-aria.

jaragunde committed:
Add clarification to hidden definition.

jaragunde commented on hidden-definition.html:
```diff
@@ -0,0 +1,2 @@
+<p>Indicates that the <a href="#dfn-element" class="internalDFN" data-link-type="dfn" id="ref-for-dfn-element-4">element</a> is not visible, <a href="#dfn-perceivable" class="internalDFN" data-link-type="dfn" id="ref-for-dfn-perceivable-3">perceivable</a>, or interactive to <em>any</em> user. An element is considered <em>hidden</em> if it or any one of its ancestor elements is not rendered or is  ...
+<p>This definition includes elements <a href="https://html.spec.whatwg.org/multipage/rendering.html#being-rendered">not being rendered</a>, which don't have an associated layout box, or rendered with a <a href="https://drafts.csswg.org/css-cascade/#computed-value" data-x-internal="computed-value">computed value</a> of <a href="https://drafts.csswg.org/css2/#propdef-visibility" data-x-internal="'v ...
```
The definition of hidden is from [WAI-ARIA](https://www.w3.org/TR/wai-aria-1.2/#dfn-hidden), I'm copying it here so we can discuss.

The second paragraph is new, it specifies the different kinds of hidden referencing other definitions and other specs too. I'm using [the innerText spec](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute) as a source for some this text, because it explains the difference between display:none and visibility:hidden in an "spec-y" way.

jaragunde replied:
Update in b2cfbb8: we discard proposing changes to WAI-ARIA definition of hidden.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,32 @@ <h4>Computation steps</h4>
-            <li id="step2A">If the <code>current node</code> is <a class="termref">hidden</a> <strong>and is not</strong> directly referenced by <code>aria-labelledby</code> or <code>aria-describedby</code>, nor directly referenced by a native host language text alternative <a class="termref">element</a> (e.g. <code>label</code> in HTML) or <a class="termref">attribute</a>, return the empty strin ...
+            <li id="step2A">If the <code>current node</code> is <a class="termref">hidden</a> <strong>and is</strong>:
+            <ul>
+              <li><strong>Not</strong> part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal, where the node directy referenced by that relation was hidden.</li>
```
This replaces the existing "current node is hidden and is not directly referenced by aria-labelledby..." text with something that explicitly includes its children.

jaragunde commented on index.html:
```diff
@@ -272,6 +293,9 @@ <h4>Computation steps</h4>
               </ul>
+              <div><details>
+                <summary>Comment:</summary>
+                <p>The result of <a href="#step2B.ii.b">step 2B.ii.b<a> in combination with <a href="#step2A">step 2A<a> means that <a class="termref">user agents</a> MUST include any hidden nodes as part of the <a class="termref">accessible name</a> or <a class="termref">accessible description</a> when traversing an <code>aria-labelledby</code> or <code>aria-describedby</code> relation and the r ...
```
This is a new comment where I hope to clarify further how UAs should behave when traversing a hidden aria-labelledby/describedby, because the actual logic is spread between 2A and 2B.

aleventhal replied:
It's also true in the <label for> case.

Another way of saying it is that the user agent MUST include *all* nodes in the subtree when the root of the relation was hidden. Correct? Maybe this is even clearer?

jaragunde replied:
It sounds clearer, my wording somehow leaves readers to wonder "what about non-hidden nodes" when there is actually no question: all are hidden and all are included.

jaragunde replied:
This was hopefully made clearer in 688456b.

jaragunde commented on index.html:
```diff
@@ -249,11 +249,32 @@ <h4>Computation steps</h4>
+            <li id="step2A">If the <code>current node</code> is <a class="termref">hidden</a> <strong>and is</strong>:
+            <ul>
+              <li><strong>Not</strong> part of an <code>aria-labelledby</code> or <code>aria-describedby</code> traversal, where the node directy referenced by that relation was hidden.</li>
+              <li><strong>Nor</strong> directly referenced by a native host language text alternative <a class="termref">element</a> (e.g. <code>label</code> in HTML) or <a class="termref">attribute</a>.</li>
```
This remains as it was, I have been "skipping" this piece of text but when examinating it closely I realized that I really didn't understand it. After some internal discussion, I'm opening a new issue about it: https://github.com/w3c/accname/issues/148

aleventhal replied:
I think we need to word this similar to the first bullet, e.g.
Not part of a native host language text traversal (e.g. label in HTML), where the root of that traversal was hidden.

jaragunde replied:
Reworked this in 94eeb38.

jaragunde committed:
Tune the examples to make them more understandable.

aleventhal commented on hidden-definition.html:
```diff
@@ -0,0 +1,2 @@
+<p>Indicates that the <a href="#dfn-element" class="internalDFN" data-link-type="dfn" id="ref-for-dfn-element-4">element</a> is not visible, <a href="#dfn-perceivable" class="internalDFN" data-link-type="dfn" id="ref-for-dfn-perceivable-3">perceivable</a>, or interactive to <em>any</em> user. An element is considered <em>hidden</em> if it or any one of its ancestor elements is not rendered or is  ...
```
The first paragraph is wrong, because aria-hidden content is often perceivable visually, when it's not also styled to be visible.
Another problem is that when an element is placed behind something to hide it, or offscreen to hide it, it's arguably still explicitly hidden.
Perhaps there is a benefit from having a table of hiding methods, and for each, describe why it matches the definition of hidden, and if not, what the reason is.
- CSS display:none, visibility:hidden, visibility: collapse, content-visibility:hidden.   Yes, explicitly hidden and not perceivable.
- CSS opacity: 0 or filter:opacity(0%), or similar SVG mechanisms: No (give reason)
- aria-hidden: Yes, explicitly hidden
- Hidden off screen or behind another object: No (give reason)

For me, having this list will help us make sure our definition matches, and is a pragmatic way to make sure implementers and authors have something easy enough to understand, and will be in alignment on the meaning. The abstract definition is just going to be too easy to misunderstand, no matter how careful we are.

jaragunde replied:
Update in b2cfbb8: added these guidelines as a "comment" in the accname spec.

jaragunde committed:
Add explicit guidelines for "hidden".

Revert proposed changes to the "hidden" definition in wai-aria and add
instead a section clarifying the specific hidden methods.

jaragunde committed:
Rework point about native host language text alternative

jaragunde committed:
Rework comment about combined behavior of 2B.ii.b and 2A

jaragunde committed:
Markup fix.

jaragunde commented:
To submit this, I think we could include a summary of what user agents will need to change to comply with this text update:

* Firefox already compliant
* Chrome would need to change the behavior for aria-hidden. Example:
  ```html
  <style>.hidden {display:none;}</style>
  <input id="el1" role="button" aria-labelledby="el2" />
  <span id="el2" class="hidden">
    <span id="el3" aria-hidden="true">hello</span>
  </span>
  ```
  Name for the input should be "hello", currently empty in Chrome.
* WebKit would need to unify behavior between different types of hidden, because it currently produces different names when using `display:none` and `visibility:hidden`. Example:
  ```html
  <style>.hidden {visibility:hidden;}</style>
  <input id="el1" role="button" aria-labelledby="el2" />
  <span id="el2" class="hidden">
    <span id="el3" class="hidden">hello</span>
  </span>
  ```
  Name for the input should be "hello", currently empty in WebKit. It would be "hello" if `display:none` was used instead.

aleventhal commented:
> To submit this, I think we could include a summary of what user agents will need to change to comply with this text update:
> 
> * Firefox already compliant

Firefox actually has some pretty hacky code around this with special cases, last time I checked. I think James Teh will tell us there is much work to be done.

jaragunde commented:
Moved here: https://github.com/w3c/accname/pull/150
